### PR TITLE
Delay the request size calculation until required by the indexing pressure framework

### PR DESCRIPTION
**Pull request by getsaurabh02**
### Description
Delay the ramBytesUsed computation for Indexing requests on the coordinator until reached a point, where absolutely necessary for IndexingPressure or ShardIndexingPressure.

### Issues Resolved
This optimises the additional request size calculation, which otherwise would have been no-op. See (#1560)

The ramBytesUsed computations are done repeatedly at multiple places (replica, primary and coordinator), and are present even in the older version of IndexingPressure. Here in OS1.2 for every request overhead would be the request size computations for the number of documents in the batch. Given that this call has neither showed up in any CPU profiling data and was neither was called out as regression when first added in the indexing path during 7.9 release, this might not be the smoking gun, however this definitely optimises it down further.

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/IndexingPressureService.java [hunk 1]
```diff
@@ -25,16 +27,18 @@ public IndexingPressureService(Settings settings, ClusterService clusterService)
         shardIndexingPressure = new ShardIndexingPressure(settings, clusterService);
     }
 
-    public Releasable markCoordinatingOperationStarted(long bytes, boolean forceExecution) {
+    public Releasable markCoordinatingOperationStarted(BulkRequest bulkRequest, boolean forceExecution) {
```
Comment (nknize) [inline server/src/main/java/org/opensearch/index/IndexingPressureService.java:30]: I see this is only used by TransportBulkAction. Is that the only place this is ever intended to be used? Shouldn't we just move this method out of the service if that's the case? If not then lets refactor the method to take the interface?

```suggestion
    public Releasable markCoordinatingOperationStarted(Accountable accountable, boolean forceExecution) {
```

File: server/src/main/java/org/opensearch/index/IndexingPressureService.java [hunk 2]
```diff
@@ -25,16 +27,18 @@ public IndexingPressureService(Settings settings, ClusterService clusterService)
...
         if (isShardIndexingPressureEnabled() == false) {
+            final long bytes = bulkRequest.ramBytesUsed();
             return shardIndexingPressure.markCoordinatingOperationStarted(bytes, forceExecution);
         } else {
             return () -> {};
         }
     }
 
-    public Releasable markCoordinatingOperationStarted(ShardId shardId, long bytes, boolean forceExecution) {
+    public Releasable markCoordinatingOperationStarted(ShardId shardId, BulkShardRequest bulkShardRequest, boolean forceExecution) {
```
Comment (nknize) [inline server/src/main/java/org/opensearch/index/IndexingPressureService.java:39]: Same here:

```suggestion
    public Releasable markCoordinatingOperationStarted(ShardId shardId, Accountable accountable, boolean forceExecution) {
```

File: server/src/main/java/org/opensearch/action/bulk/TransportBulkAction.java [hunk 3]
```diff
@@ -631,7 +630,7 @@ protected void doRun() {
                 final boolean isOnlySystem = isOnlySystem(bulkRequest, clusterService.state().metadata().getIndicesLookup(), systemIndices);
```
Comment (nknize) [inline server/src/main/java/org/opensearch/action/bulk/TransportBulkAction.java:630]: Can we delay this `isOnlySystem` call as well? Like the `ramBytesUsed()` it's not needed for anything other than the releasable.

**Review by nknize — CHANGES_REQUESTED** [review]
Love the change! I have a few questions based on the minimal usage of this method. We may consider refactoring to make this a bit more straightforward.

I think we should also take this opportunity to add javadocs for future contributors and debugging.

File: server/src/main/java/org/opensearch/index/IndexingPressureService.java (hunk 1 above)
Comment (getsaurabh02) [reply server/src/main/java/org/opensearch/index/IndexingPressureService.java]: Yes, it makes perfect sense to use an Interface instead and keep the contract clean. Refactored to use a Supplier.

File: server/src/main/java/org/opensearch/index/IndexingPressureService.java (hunk 2 above)
Comment (getsaurabh02) [reply server/src/main/java/org/opensearch/index/IndexingPressureService.java]: same as above.

File: server/src/main/java/org/opensearch/action/bulk/TransportBulkAction.java (hunk 3 above)
Comment (getsaurabh02) [reply server/src/main/java/org/opensearch/action/bulk/TransportBulkAction.java:630]: Keeping this consistent with the other interface for Node level accounting ``markCoordinatingOperationStarted(LongSupplier bytes, boolean forceExecution)`` as it is supplied with a precomputed boolean, being used elsewhere too. 

Also given ``isOnlySystem`` streams over a precomputed collection (unmodifiable set) of different indices being updated in that bulk request, we don’t expect this number (indices in a single request) to grow too high to be an overhead.

**Comment by getsaurabh02** [comment]
Thank you @nknize for taking the look. I have updated the PR and added some java doc too.

**Comment by nknize** [comment]
stale PR. Reinvoking gradle check

**Comment by nknize** [comment]
start gradle check

**Review by dblock — APPROVED** [review]
@nknize you're ok with this?

**Pull request merged**